% Template for PLoS
% Version 1.0 January 2009
%
\documentclass[10pt]{article}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{article}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **


%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW

%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Genomer - a Swiss army knife for genome scaffolding.}
}
% Insert Author names, affiliations and corresponding author email.
\\
Michael D. Barton$^{1\ast}$, 
Hazel A. Barton$^{1}$
\\
\bf{1} Biology Department, The University of Akron, Akron, OH, 44325-3908, USA
\\
$\ast$ E-mail: mail@michaelbarton.me.uk
\end{flushleft}

\section*{Abstract}

The increasing accessibility and reduced costs of sequencing has made genome
analysis accessible to more and more researchers. In contrast there remains a
steep learning curve in the subsequent computational steps required to process
raw sequence reads into a database-deposited genome sequence. Here we describe
``Genomer,'' a tool to simplify the manual tasks of finishing and uploading a
genome sequence to a database. Genomer can format a genome scaffold into the
common files required for submission to GenBank. Genomer simplifies updating a
genome scaffold by allowing a human-readable YAML format file to be edited
instead of large sequence files. Genomer is written as a command line tool and
is an effort to make the manual, and increasingly common, process of genome
scaffolding more robust and reproducible. Extensive documentation and video
tutorials are available at http://next.gs.

\section*{Introduction}

The decreasing costs and increasing diversity of high-throughput sequencing
methods is making genome analysis a common method to tackle unresolved
questions in microbiology \cite{loman2012b}. Sequencing can produce
$10e^{6}$-$10e^{10}$ short nucleotide reads, which must be assembled into
larger contigs then scaffolded into larger megabase fragments or complete
chromosomes. These larger DNA sequences are then annotated with genomic
features such as protein-coding genes. Each of these required steps can however
produce imperfect results and remains the subject of active research
\cite{earl2011,quail2012,beckloff2012}. Therefore manual curation of a genome
sequence and corresponding annotations may still improve upon the results of
automated methods.

Manually editing a genome sequence or annotation is a non-trivial problem
requiring effort on a researcher's part. The large size of FASTA sequence files
or general feature format (GFF) annotation files makes changing these in a text
editor difficult. Tasks such as organising project files, adding additional
sequences to an assembly, updating incorrect annotations or preparing files for
database deposition all divert effort from the main goal of analysing a genome.
As a result genome projects have a steep learning curve for researchers
embarking on genome sequencing for the first time, while still requiring effort
for experienced researchers.

We anticipate these problems have resulted in individual groups independently
producing in-house programming scripts to automate repetitive tasks. An example
of such tasks include finding protein names that do not match a convention of
three lower case characters followed by one upper case character (abcD). Such
tasks are manually laborious but simple to automate computationally. Individual
research groups producing their own scripts for such tasks however leads to a
repetition of effort. In contrast open-source projects can lead to a pooling of
community effort resulting in a higher standard of software and preventing
reinvention of the same code \cite{ince2012}.

During the process of sequencing several \emph{Pseudomonas fluorescens} strains
we developed a tool, called ``Genomer'', to ease and simplify many of the
common steps in our genome project. Several graphical user interface tools are
available to work with genome projects \cite{tanaka2006, wilkinson2002,
lopez2011, carver2012, gordon2003}. Genomer instead provides a simple, command
line interface for genome finishing allowing scriptable automation and
reproducible results. We have open-sourced this tool under the MIT license and
here we describe its application in a microbial genome project.

\section*{Implementation}

Genomer is written in the Ruby programming language \cite{ruby-lang,goto2010}
and tested against the 1.8.7 and 1.9.3 versions. Genomer is built upon the
Rubygems and Bundler Ruby package management libraries. Rubygems is a software
management system for Ruby which allows automatic downloading and installation
of third-party software libraries from the rubygems.org website. Bundler is
software for determining which specific versions of the available third
libraries should be used.

Genomer uses these two libraries to create a plugin system where third-party
software can be written and included. Plugin creators need only to prefix the
name of their plugin with \verb+genomer-plugin-+ and upload this to
rubygems.org or make the plugin available as a public git repository. A plugin
is specified within a Genomer project in a file named `Gemfile.' The
\verb+bundle update+ command will then automatically download and install the
required plugins.

Genomer is implemented as a command line tool and tested on both Mac OS X and
Linux systems. The latest version of genomer can installed from the Rubygems
package management system with the command \verb+gem install genomer+. Genomer
is based on our previously described ``scaffolder'' software and is invoked on
the command line with the command \verb+genomer+. Genomer provides extensive
documentation via UNIX manual pages and on the genomer website at
http://next.gs.

\section*{Results and Discussion}

Genomer was developed during our \emph{P. fluorescens} genome projects to
automate and simplify the manual steps required when finishing a microbial
genome. Genomer provides the following functionality to facilitate this:

\begin{description} 

  \item[Simple editing of a draft genome sequence.]{Genomer is built on the
  existing Scaffolder \cite{barton2012} file format for assembling draft genome
  sequences. This is format requires only the order and IDs of each contig be
  specified. This thereby simplifies the process of re-organising and trimming
  contigs in draft genome without having to edit large nucleotide sequences
  manually.}

  \item[Mapping of annotations onto the assembled sequence.]{Genomer maps the
  coordinates of contig annotations to respective positions in the draft genome
  sequence. This allows the scaffold file to be continuously improved and
  updated even after annotation. For instance, additional PCR sequences can be
  used to close gaps in the assembly without reannotation or manually updating
  annotation coordinates.}

  \item[Generation of files for submission to GenBank.]{Submitting a genome
  sequence to GenBank database requires generating specific files. These may
  include a FASTA file of the draft sequence, a table of annotations, a FASTA
  file containing the individual contig sequences, and `a golden path' (AGP)
  \cite{agp-spec} file describing the placement of contigs. Genomer automates
  the generation of all these files from the scaffold file and corresponding
  GFF3 file.}

  \item[A stable interface and streamlined install process]{Genomer provides a
  plugin system built using the existing Ruby package management system:
  RubyGems and Bundler. This allows genomer and its plugins to be automatically
  installed without requiring manual downloading and compilation by the user.
  This eliminates a common problem in bioinformatics where software may require
  technical expertise to first compile and install before being used. Plugins
  can also be locked to specific versions to prevent backwards incompatible
  software changes breaking an existing workflow.}

  \item[Integration with the command line.]{Genomer is built as a command line
  tool built around plain text files. This allows for integration with common
  Linux tools tools such as GNU Make or git. This thereby allow reproducible
  scripting of the genome finishing process.}

\end{description}

Genomer is written for the command-line to use in shells such as bash or zsh.
Command line tools are simple to automate using scripting, allowing
reproducible genome finishing and sharing genome projects between researchers.
For instance, during our microbial genome project we automated Genomer using
GNU Make build files, allowing the finishing steps to be repeated automatically
using the \verb+make+ command. This is an advantage of command-line tools
versus GUI-based tools, the latter are harder to automate and share steps.
Therefore Genomer may be of particular interest to bioinformaticians who prefer
automated approaches and scripting on the command line.

Additionally as Genomer uses plain text it is easy to store versions of the
project using a revision control system such as git \cite{git-scm}. This allows
tracking changes to the project and reverting errors back to earlier versions.
Example build files from our genome projects can be found on GitHub for a
simple plasmid \cite{plasmid-github} and a more complex genome
\cite{genome-github} sequence.

We have released Genomer as open source on GitHub \cite{genomer-github} and
have prepared documentation and video tutorials at http://next.gs. As we have
used this tool extensively to simplify our own genome projects we believe
genomer may also be useful to others in the field. We will illustrate the
possible application of genomer with an example use case.

\subsection*{Genomer use case}

\begin{enumerate}

  \item{A Genomer project is organised around a set of already assembled and
  annotated contigs. For example we assemble our reads into contigs using the
  A5 pipeline \cite{tritt2012} and and then subsequently annotate them using
  the Integrated Microbial Genomes resource \cite{markowitz2006}. We then use
  these contigs as a starting point to build a draft genome sequence in
  Genomer.}

  \item{A genome scaffold is written in the scaffolder file format
  \cite{barton2012}. This specifies the order and orientation of contigs in the
  scaffold. Any unresolved gap regions can also be specified in this file.
  Paired-end sequencing or comparison to reference genomes are a source of
  contig orientation and order. In our genome projects we determine contig
  order by aligning contigs to a reference genome using nucmer
  \cite{kurtz2004}. A detailed overview of the scaffolder file format can be
  found on the genomer website.}

  \item{The \verb+genomer view+ command can be used to generate files of the
  assembled scaffold. These may include the FASTA file of the assembly or a
  GFF3 file of updated annotations corresponding to the assembled genome
  sequence. These files can then be used for downstream analyses of the
  sequence.}

  \item{The genome scaffold can be continuously updated through closing any
  remaining gaps. During our microbe sequencing projects we closed gaps through
  a combination of Sanger PCR and \emph{in silico} analysis. Changes to the
  scaffold file automatically propagate updates in the generated files when the
  \verb+genomer view+ command is rerun.}

  \item{When researchers are satisfied with the status of their draft the
  required files for submission to GenBank can be generated: assembled FASTA,
  contig FASTA, annotation tables, and scaffold layout in AGP. Additional
  options are used to add the required prefix to gene annotation IDs and begin
  locus ID numbering at the origin of the sequence. Genomer does not produce
  the .asn file which must be submitted to GenBank, but instead the files that
  may be used as the input to tbl2asn or sequin. }

\end{enumerate}

\subsection*{Limitations}

Genomer was written to satisfy our needs for building and finishing a draft
microbial genomes less than 10MBp in size. Draft sequence assembly takes place
in memory and therefore building very large sequences in memory may limit
performance. We point out however that sizes of even 3GBp should fit in the
memory of a modern laptop or desktop computer.

Genomer simplifies the process of generating the annotation table file required
for submission to GenBank where only gene entries are required. Specifically a
user need only provide a GFF3 file of annotations and Genomer can automatically
generate the additional encoded protein features. This however assumes no
intron/exon structure and therefore users wishing to submit annotations with
alternative splicing cannot take advantage of this genomer feature.

\subsection*{Summary}

Genomer is an open-source tool to simplify and automate repetitive and
time-consuming tasks required when finishing a microbial genome project. This
software is available on GitHub \cite{genomer-github} with documentation and
video tutorials on the http://next.gs website. This tool has been useful in our
own genome projects and we believe it will also be useful to other researchers
especially smaller research groups entering the field for the first time.

% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgements}

This work was supported by the National Institute for Health: IDeA Network of
Biomedical Research Excellence (KY-INBRE) grant (NIH 2P20 RR016481-09) and the
NIH R15 AREA Program grant (R15GM079775).

%\section*{References}
% The bibtex filename
\bibliography{article}

\end{document}
